doc/sphinx/language/core/definitions.rst

```diff
@@ -1,5 +1,20 @@
-Definitions
-===========
+Definitions and theorems
+========================
+
+A definition can be seen as a way to give a meaning to a name or as a
+way to abbreviate a term of some type. The name can later
+be replaced at any time by its definition.
+
+On its side, a theorem is a statement with a proof. One can view
+the name of a theorem as a way to abbreviate the given proof, in the
+same way as the name of a definition abbreviates a term. That is, in
+the case of definitions (and related forms such as :cmd:`Fixpoint` or
+:cmd:`CoFixpoint`), the term is the body of the definition and the
+type is the type of the body. In the case of a theorem, lemma,
+corollary, etc. the term is the proof and the type is the statement.
```

Comment:
> What do you mean by "References to the theorem may use its term or type, depending on the context."? Otherwise ok.

What I meant was if you `apply` the theorem, the type of the theorem is used.  `Print theorem` will show both.  Maybe my thinking is still fuzzy, feel free to drop the sentence.

> I think it is also worth to insist somewhere that Definition and Theorem are just two different views at the same kind of syntactic object.

Yes, a good point.  I tend to forget the case where the user has to provide a proof for the Definition.

Reply:
I hope you'll at least use the first three sentences I suggest here.  IMO short and sweet.

Reply:
> What I meant was if you `apply` the theorem, the type of the theorem is used.

The proof also is also used, in some sense. If your theorem is, say, `{x : nat & x > 1}` and you use this statement to eventually compute an actual `nat` bigger than 1, this `nat` is found in the proof that you applied (and it can be 2, or 3, etc. depending on the proof you give).

> Maybe my thinking is still fuzzy, feel free to drop the sentence.

I would say that it is the standard usage which is ambiguous. Syntactically, there is no difference between a proof of some statement and a term of some type, that is when you apply a theorem, what you syntactically do is to apply the proof. However, in the case of a pair proof/statement, the focus is on the statement and we usually intend that the exact proof does not matter (and thus use `Qed` to enforce that), while in the case of a pair term/type, the focus is on the term and the type is intended to be an auxiliary re-derivable information. (Don't know if this explanaton helps.)

Reply:
> Syntactically, there is no difference between a proof of some statement and a term of some type, that is when you apply a theorem, what you syntactically do is to apply the proof.

When you apply a theorem, the theorem name is inserted into the proof term of Show Proof.  I expect Qed doesn't later recheck the referenced theorem--right?